File: src/main/java/legacy/kms/chapter02/discount/DiscountConditionSet.java
package legacy.kms.chapter02.discount;

import legacy.kms.chapter02.Screening;

import java.util.Arrays;
import java.util.List;

// 등록된 모든 조건을 만족해야한다
public class DiscountConditionSet implements DiscountCondition {

    private List<DiscountCondition> conditions;

    DiscountConditionSet(DiscountCondition ... conditions) {
        this.conditions = Arrays.asList(conditions);
    }

    @Override
    public boolean isSatisfiedBy(Screening screening) {
        boolean isSatisfied = true;
        for(DiscountCondition condition : conditions){
            if(!condition.isSatisfiedBy(screening)){
                isSatisfied = false;
                break;
            }
        }
        return isSatisfied;
    }
}

File: src/main/kotlin/kms/chapter02/discount/DiscountConditionSet.kt
package kms.chapter02.discount

import kms.chapter02.Screening

internal class DiscountConditionSet(
    private val conditions: List<DiscountCondition>,
) : DiscountCondition {

    constructor(vararg conditions: DiscountCondition) : this(conditions.toList())

    override fun isSatisfiedBy(screening: Screening): Boolean {
        return conditions.all { it.isSatisfiedBy(screening) }
    }
}
